reports-scheduler/src/main/java/com/amazon/opendistroforelasticsearch/reportsscheduler/job/ReportsSchedulerJobRunner.java

```diff
@@ -0,0 +1,139 @@
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amazon.opendistroforelasticsearch.reportsscheduler.job;
+
+import static com.amazon.opendistroforelasticsearch.reportsscheduler.common.Constants.JOB_QUEUE_INDEX_NAME;
+
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.index.IndexRequest;
+import org.elasticsearch.action.index.IndexResponse;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.plugins.Plugin;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import com.amazon.opendistroforelasticsearch.jobscheduler.spi.JobExecutionContext;
+import com.amazon.opendistroforelasticsearch.jobscheduler.spi.ScheduledJobParameter;
+import com.amazon.opendistroforelasticsearch.jobscheduler.spi.ScheduledJobRunner;
+import com.amazon.opendistroforelasticsearch.reportsscheduler.job.parameter.JobParameter;
+
+/**
+ * Reports scheduler job runner class.
+ *
+ * <p>The job runner should be a singleton class if it uses Elasticsearch client or other objects
+ * passed from Elasticsearch. Because when registering the job runner to JobScheduler plugin,
+ * Elasticsearch has not invoke plugins' createComponents() method. That is saying the plugin is not
+ * completely initalized, and the Elasticsearch {@link org.elasticsearch.client.Client}, {@link
+ * ClusterService} and other objects are not available to plugin and this job runner.
+ *
+ * <p>So we have to move this job runner intialization to {@link Plugin} createComponents() method,
+ * and using singleton job runner to ensure we register a usable job runner instance to JobScheduler
+ * plugin.
+ *
+ * <p>This reports scheduler job runner will write the report_definition_id
+ */
+public class ReportsSchedulerJobRunner implements ScheduledJobRunner {
+  private static final Logger log = LogManager.getLogger(ScheduledJobRunner.class);
+  private static ReportsSchedulerJobRunner INSTANCE;
+
+  private ClusterService clusterService;
+  private ThreadPool threadPool;
+  private Client client;
+
+  public static ReportsSchedulerJobRunner getJobRunnerInstance() {
+    if (INSTANCE != null) {
+      return INSTANCE;
+    }
+    synchronized (ReportsSchedulerJobRunner.class) {
+      if (INSTANCE != null) {
+        return INSTANCE;
+      }
+      INSTANCE = new ReportsSchedulerJobRunner();
+      return INSTANCE;
+    }
+  }
+
+  private ReportsSchedulerJobRunner() {
+    // Singleton class, use getJobRunner method instead of constructor
+  }
+
+  public void setClient(Client client) {
```

Comment:
Why is this class mutable (having setters)?

Reply:
The job runner should be a singleton class if it uses Elasticsearch client or other objects passed from Elasticsearch. Because when registering the job runner to JobScheduler plugin, Elasticsearch has not invoke plugins' createComponents() method. That is saying the plugin is notcompletely initialized, and the Elasticsearch client and clusterService and other objects are not available to plugin and this job runner.

So we have to move this job runner initialization to `createComponents()` method in reportsSchedulerPlugin, and using singleton job runner to ensure we register a usable job runner instance to JobScheduler plugin.

Tried to declare it as final, and create the instance using constructor in `createdComponents()`and there is always exception thrown saying `clusterService has not been initialized`

Reply:
May be in that case I think that better way is to not create instance in getJobRunnerInstance() but only return if it exists. createComponents() will create object with proper initialization. anyone calling getJobRunnerInstance() before that will get InvalidState exception. with that we can assure immutability?